File: scripts/build-capture-candidate.mjs
// Separate development artifact. Never changes extension/dist or package selection.
import { mkdir, lstat, readFile, writeFile } from "node:fs/promises";
import { createHash } from "node:crypto";
import { execFileSync } from "node:child_process";
import path from "node:path";
import { extensionLicenses } from "./extension-licenses.mjs";
import { candidatePayloads, validateCaptureCandidate } from "./capture-candidate-policy.mjs";

const output = path.resolve(process.argv[2] ?? ".");
const relative = path.relative(path.resolve("artifacts"), output);
if (
  process.argv.length !== 3 ||
  !relative ||
  relative.startsWith("..") ||
  path.isAbsolute(relative)
)
  throw new Error("A new directory beneath artifacts is required");
for (let parent = path.dirname(output); ; parent = path.dirname(parent)) {
  const stat = await lstat(parent);
  if (!stat.isDirectory() || stat.isSymbolicLink())
    throw new Error("Candidate parent is not ordinary");
  if (parent === path.dirname(parent)) break;
}
const git = (...args) =>
  execFileSync("git", args, { encoding: "utf8", maxBuffer: 1024 * 1024, windowsHide: true });
const revision = git("rev-parse", "HEAD").trim();
if (!/^[a-f0-9]{40}$/u.test(revision)) throw new Error("Candidate revision unavailable");
const initiallyDirty = git("status", "--porcelain").length !== 0;
const manifest = JSON.parse(await readFile("extension/candidate/manifest.json", "utf8"));
validateCaptureCandidate(manifest);
await mkdir(output); // Exclusive; failed builds are preserved, never adopted.
process.env.ESBUILD_WORKER_THREADS = "0";
process.env.ESBUILD_MAX_BUFFER = "16777216";
delete process.env.ESBUILD_BINARY_PATH;
const { buildSync } = await import("esbuild");
const payloads = {};
for (const [name, entry] of Object.entries({
  "background.js": "extension/src/automatic-background.ts",
  "click.js": "extension/src/capture-click.ts",
  "manager.js": "extension/src/manager.ts",
})) {
  const result = buildSync({
    bundle: true,
    entryPoints: [entry],
    format: "iife",
    write: false,
    legalComments: "none",
    platform: "browser",
    target: "firefox156",
    sourcemap: false,
  });
  if (result.outputFiles.length !== 1) throw new Error("Candidate bundle shape refused");
  payloads[name] = result.outputFiles[0].contents;
}
payloads["manifest.json"] = Buffer.from(JSON.stringify(manifest));
for (const name of ["manager.html", "manager.css"])
  payloads[name] = await readFile(`extension/src/${name}`);
for (const [name, text] of Object.entries(await extensionLicenses()))
  payloads[name] = Buffer.from(text);
if (Object.keys(payloads).sort().join() !== candidatePayloads.slice().sort().join())
  throw new Error("Candidate inventory refused");
const hashes = {};
for (const [name, data] of Object.entries(payloads)) {
  if (!data.length || data.length > 1024 * 1024)
    throw new Error("Candidate payload bound exceeded");
  await writeFile(path.join(output, name), data, { flag: "wx" });
  hashes[name] = createHash("sha256").update(data).digest("hex");
}
if (git("rev-parse", "HEAD").trim() !== revision)
  throw new Error("Source revision changed during build");
await writeFile(
  path.join(output, "BUILD.json"),
  JSON.stringify({
    version: 1,
    candidate: true,
    qualification: false,
    source_commit: revision,
    source_dirty: initiallyDirty || git("status", "--porcelain").length !== 0,
    files: hashes,
  }),
  { flag: "wx" },
);

File: scripts/build-capture-candidate.py
"""Build a separately labelled unsigned capture XPI; no browser or installation actions."""
import argparse
import hashlib
import json
import re
from pathlib import Path
import subprocess
import time
import zipfile

ROOT = Path(__file__).resolve().parents[1]
PAYLOADS = {"background.js", "click.js", "manager.js", "manifest.json", "manager.html", "manager.css", "LICENSE.txt", "THIRD-PARTY-NOTICES.txt"}


def unique(pairs):
    result = {}
    for key, value in pairs:
        if key in result: raise RuntimeError("duplicate candidate field")
        result[key] = value
    return result


def build(output):
    output = Path(output).absolute()
    # Node independently enforces the new artifacts-only directory and ordinary parents.
    child = subprocess.Popen(["node", str(ROOT / "scripts/build-capture-candidate.mjs"), str(output)],
                             cwd=ROOT, stdin=subprocess.DEVNULL, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    try:
        code = child.wait(timeout=60)
    except BaseException:
        # Never kill the build parent and abandon its synchronous compiler child.
        while True:
            try: child.wait(); break
            except BaseException:
                try: time.sleep(.1)
                except BaseException: pass
        raise
    if code != 0: raise RuntimeError("capture candidate build refused; preserve output")
    with (output / "BUILD.json").open("rb") as source: raw = source.read(16385)
    if len(raw) > 16384: raise RuntimeError("candidate metadata bound exceeded")
    info = json.loads(raw, object_pairs_hook=unique)
    if (not isinstance(info, dict) or set(info) != {"version", "candidate", "qualification", "source_commit", "source_dirty", "files"}
            or type(info.get("version")) is not int or not isinstance(info.get("source_commit"), str)
            or not re.fullmatch(r"[a-f0-9]{40}", info["source_commit"])
            or info.get("candidate") is not True or info.get("qualification") is not False
            or type(info.get("source_dirty")) is not bool or info.get("version") != 1
            or set(info.get("files", {})) != PAYLOADS):
        raise RuntimeError("candidate metadata refused")
    payloads = {}
    for name in sorted(PAYLOADS):
        file = output / name
        if file.is_symlink() or file.is_junction() or not file.is_file(): raise RuntimeError("candidate file refused")
        with file.open("rb") as source: data = source.read(1024 * 1024 + 1)
        if not 0 < len(data) <= 1024 * 1024 or hashlib.sha256(data).hexdigest() != info["files"][name]:
            raise RuntimeError("candidate payload differs")
        payloads[name] = data
    manifest = json.loads(payloads["manifest.json"], object_pairs_hook=unique)
    if (manifest.get("version") != "0.2.0" or manifest.get("incognito") != "not_allowed"
            or manifest.get("permissions") != ["nativeMessaging", "menus", "storage", "webRequest", "webRequestBlocking"]
            or manifest.get("host_permissions") != ["http://*/*", "https://*/*"]
            or manifest.get("browser_specific_settings", {}).get("gecko", {}).get("id") != "download-manager@halcyonxp.local"):
        raise RuntimeError("candidate authority differs")
    xpi = output / "download-manager-capture-candidate.xpi"
    with zipfile.ZipFile(xpi, "x", compression=zipfile.ZIP_STORED) as archive:
        for name, data in payloads.items(): archive.writestr(name, data)
    # Independent archive readback before recording its identity.
    with zipfile.ZipFile(xpi) as archive:
        if set(archive.namelist()) != PAYLOADS or len(archive.namelist()) != len(PAYLOADS): raise RuntimeError("candidate archive inventory differs")
        for name, data in payloads.items():
            if archive.read(name) != data: raise RuntimeError("candidate archive bytes differ")
    result = {"candidate": True, "qualification": False, "source_commit": info["source_commit"],
              "source_dirty": info["source_dirty"], "xpi_sha256": hashlib.sha256(xpi.read_bytes()).hexdigest()}
    with (output / "candidate.json").open("x", encoding="utf-8") as target: json.dump(result, target, indent=2)
    return result


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    print(json.dumps(build(args.output)))
